source/libraries/HelperFunctions.ts

```diff
@@ -71,3 +74,11 @@ export async function initializeTestRpcClientOptions(ethHttpProviderPort: number
 
     return { gasLimit: GAS_BLOCK_AMOUNT, accounts: accountOptions };
 }
+
+export async function waitForTransactionToBeSealed(ethjsQuery: EthjsQuery, transactionHash: string): Promise<void> {
+    let transaction = await ethjsQuery.getTransactionByHash(transactionHash);
+    while (transaction.blockNumber == null) {
+        await sleep.sleep(5);
+        transaction = await ethjsQuery.getTransactionByHash(transactionHash);
```

Comment:
It seems like the common case is that we'll want the transaction receipt after the transaction shows up in the block.  Consider using `ethjsQuery.getTransactionReceipt(transactionHash)` instead (which also includes block number and block hash) and returning it when done.  This will save the caller from having to then do an extra round trip afterwards to get the receipt.

Reply:
Using `getTransactionReceipt` (without first checking `getTransactionByHash`) was partially responsible for this problem, because Parity and geth make the transaction receipt available at different times, and when geth doesn't make the transaction receipt available it simply returns `null`.  `getTransactionByHash` behaves the same way in both clients AFAICT (and is also the way that ethrpc checks for this FWIW).

Reply:
Couldn't we just do this to address the "geth doesn't make receipts available until transaction is mined"?
```
while (!receipt || !receipt.blockNumber)
```

Reply:
IMO we should use the lookup (`getTransactionByHash`) that we've tested extensively and that we know works.  The receipts data structure (within the ethereum node) is different than the transactions data structure and it appears to have some strange bugs/edge cases, see for example https://github.com/ethereum/go-ethereum/issues/1746

Reply:
@MicahZoltu I tried changing getTransactionByHash to getTransactionReceipt to see what would happen.  Here is the result in geth:
```
{ Error: [ethjs-rpc]  error with payload {"id":6350900769400,"jsonrpc":"2.0","params":["0x3be152e828c8321a73267d3528e00a86f5d4a27f22ea3b754c85c7abf5e5204f"],"method":"eth_getTransactionReceipt"} Error: [ethjs-provider-http] Invalid JSON RPC response from provider
    host: http://localhost:8545
    response:  ""
    responseURL: undefined
    status: 0
    statusText: 
  
    at /home/jack/src/augur-core/node_modules/ethjs-rpc/lib/index.js:53:26
    at XMLHttpRequest.request.onreadystatechange (/home/jack/src/augur-core/node_modules/ethjs-provider-http/lib/index.js:91:7)
    at XMLHttpRequestEventTarget.dispatchEvent (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:64:18)
    at XMLHttpRequest._setReadyState (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:353:12)
    at XMLHttpRequest._onHttpRequestError (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:542:12)
    at ClientRequest.<anonymous> (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:413:24)
    at emitOne (events.js:115:13)
    at ClientRequest.emit (events.js:210:7)
    at Socket.socketErrorListener (_http_client.js:401:9)
    at emitOne (events.js:115:13)
  value: 
   { Error: [ethjs-provider-http] Invalid JSON RPC response from provider
    host: http://localhost:8545
    response:  ""
    responseURL: undefined
    status: 0
    statusText: 
  
    at invalidResponseError (/home/jack/src/augur-core/node_modules/ethjs-provider-http/lib/index.js:43:23)
    at XMLHttpRequest.request.onreadystatechange (/home/jack/src/augur-core/node_modules/ethjs-provider-http/lib/index.js:88:17)
    at XMLHttpRequestEventTarget.dispatchEvent (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:64:18)
    at XMLHttpRequest._setReadyState (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:353:12)
    at XMLHttpRequest._onHttpRequestError (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:542:12)
    at ClientRequest.<anonymous> (/home/jack/src/augur-core/node_modules/xhr2/lib/xhr2.js:413:24)
    at emitOne (events.js:115:13)
    at ClientRequest.emit (events.js:210:7)
    at Socket.socketErrorListener (_http_client.js:401:9)
    at emitOne (events.js:115:13)
     value: 
      XMLHttpRequest {
        onloadstart: null,
        onprogress: null,
        onabort: null,
        onerror: null,
        onload: null,
        ontimeout: [Function],
        onloadend: null,
        _listeners: {},
        onreadystatechange: [Function],
        _anonymous: undefined,
        readyState: 4,
        response: null,
        responseText: '',
        responseType: '',
        status: 0,
        statusText: '',
        timeout: 0,
        upload: [Object],
        _method: 'POST',
        _url: [Object],
        _sync: false,
        _headers: [Object],
        _loweredHeaders: [Object],
        _mimeOverride: null,
        _request: null,
        _response: null,
        _responseParts: null,
        _responseHeaders: null,
        _aborting: null,
        _error: null,
        _loadedBytes: 0,
        _totalBytes: 0,
        _lengthComputable: false } } }
```

Reply:
As of Byzantium we _have_ to look at the receipt so we can know whether the transaction succeeded or failed.  The GitHub issue you linked is about chain reorgs which is actually a much harder problem to solve and using `getTransactionByHash` doesn't solve it either.  Are there other "known issues" with transaction receipts that we need to worry about?

At the moment, this script assumes that there are no chain reorgs and that transactions do not fail.  Neither of these are actually safe assumptions and we should weigh whether or not we want to devote development effort to dealing with them in the deployment scripts.  I think we _should_ come up with a solution for transaction failure since its pretty easy (see below).  I'm not sure what to do about chain reorgs as that is a much harder problem to solve.

```
// TODO: make this not an `async` function because Node is stupid and will log a warning about unhandled promises
async function exitOnTransactionFailure(transactionHash: string) {
  const receipt = await waitForTransactionToSeal(transactionHash);
  if (receipt.status) return;
  console.log(`Transaction failed: ${transactionHash}`);
  process.exit(1);
}
```

Reply:
@tinybike Hmm, that error suggests that your Geth node isn't actually responding to you, not that Geth returned an error.  Are you sure Geth was on/live at the time of the call?  I would have expected a null JSON-RPC response, not Geth to simply reject your HTTP connection (that isn't even an HTTP error, that is a TCP connection failure error).

Reply:
Right, the receipt tells you if the transaction succeeded or failed.  This is an extra and quite valuable piece of info.  `getTransactionByHash` just tells you whether or not the transaction has been _sealed_.  There are definitely pathological cases the current simple implementation won't catch, for example if a transaction is silently dropped from the txpool and needs to be re-sent, or (like you said) if a transaction fails, if a chain reorg happens, etc.  However `getTransactionByHash` by itself without other checks should make this work most of the time!  `getTransactionReceipt` should (as you pointed out) accomplish the same thing but it appears to have some sharp edges so IMO for now it's sensible to just use `getTransactionByHash`.

Reply:
> @tinybike Hmm, that error suggests that your Geth node isn't actually responding to you, not that Geth returned an error. Are you sure Geth was on/live at the time of the call? I would have expected a null JSON-RPC response, not Geth to simply reject your HTTP connection (that isn't even an HTTP error, that is a TCP connection failure error).
Digging a bit.  That error was thrown when I had the sleep time set to 50ms.  Just re-tried it (w/ 50ms sleep time) and the error is _also_ thrown when `getTransactionByHash` is used instead of `getTransactionReceipt`.  My guess is what's happening is just that geth's RPC server is overloading and sending back empty (??) responses.  I would have thought this would be a `'0x'` rather than a `''` response though.

Reply:
The error is a TCP error, not an HTTP or JSON-RPC error.  You can tell by the `status: 0` bit (that is HTTP status codes).  No idea why Geth is completely tanking like this.

Reply:
Another option would be an exponential backoff.  This would make it so against insta-seal or short block time chains we get the response quickly, and against long block time chains we don't hammer the server.  I'm thinking 10 ms to 5 seconds exponential (2*prev) backoff would be good?